scripts/usb_reset.py

```diff
@@ -14,46 +14,285 @@
+    detach.add_argument("-i", dest="id", type=int, nargs=2, required=True,
+                        help="restore ownership to uid gid")
+
+    cleanup = subparsers.add_parser("cleanup", help="clean up chroot directory")
+    cleanup.add_argument("-d", dest="domid", type=int, required=True,
+                         help="specify the domid of the VM")
+
+    return parser.parse_args()
+
 
-#USBDEVFS_RESET _IO('U', 20)
-USBDEVFS_RESET = (ord('U') << 8) | 20
+def get_root_dir(domid):
+    return "/var/xen/qemu/root-{}".format(domid)
 
 
+def get_cg_dir(domid):
+    return "/sys/fs/cgroup/devices/qemu-{}".format(domid)
+
+
+# throw IOError, ValueError
 def read_int(path):
     with open(path) as f:
         return int(f.readline())
 
 
-def reset_device(bus, dev):
-    with open("/dev/bus/usb/{0:03d}/{1:03d}".format(bus, dev), "w") as f:
-        fcntl.ioctl(f.fileno(), USBDEVFS_RESET, 0)
+def dev_path(device):
+    # check device node pattern
+    # example: "4-1", "1-2.1.2"
+    pat = re.compile(r"\d+-\d+(\.\d+)*$")
+    if pat.match(device) is None:
+        log.error("Unexpected device node: {}".format(device))
+        exit(1)
+    try:
+        bus = read_int("/sys/bus/usb/devices/{}/busnum".format(device))
+        dev = read_int("/sys/bus/usb/devices/{}/devnum".format(device))
+        return "/dev/bus/usb/{0:03d}/{1:03d}".format(bus, dev)
+    except (IOError, ValueError) as e:
+        log.error("Failed to get device path {}: {}".format(device, str(e)))
+        exit(1)
 
 
-if __name__ == "__main__":
-    log.logToSyslog(level=logging.DEBUG)
+def get_ctl(path, mode):
+    """get the string to control device access for cgroup
+    :param path: the device file path
+    :param mode: either "r" or "rw"
+    :return: the string to control device access
+    """
+    try:
+        st = os.stat(path)
+    except OSError as e:
+        log.error("Failed to get stat of {}: {}".format(path, str(e)))
+        raise
+
+    t = ""
+    if S_ISBLK(st.st_mode):
+        t = "b"
+    elif S_ISCHR(st.st_mode):
+        t = "c"
+    if t and mode in ("r", "rw"):
+        return "{} {}:{} {}".format(t, os.major(st.st_rdev), os.minor(
+            st.st_rdev), mode)
+    raise RuntimeError("Failed to get control string of {}".format(path))
 
-    if len(sys.argv) != 2:
-        sys.exit("usage: {} device_node".format(sys.argv[0]))
 
-    device = sys.argv[1]
-    pattern = re.compile(r"^\d+-\d+(\.\d+)*$")
-    if pattern.match(device) is None:
-        log.debug("unexpected device node: {}".format(device))
+def _device_ctl(path, domid, allow):
+    cg_dir = get_cg_dir(domid)
+    file_name = "/devices.allow" if allow else "/devices.deny"
+    try:
+        with open(cg_dir + file_name, "w") as f:
+            f.write(get_ctl(path, "rw"))
+    except (IOError, OSError, RuntimeError) as e:
+        log.error("Failed to {} {}: {}".format(
+            "allow" if allow else "deny", path, str(e)))
         exit(1)
 
```

Comment:
Consider xapi might call detach after cleanup,
in which case, the cgroup dir is already removed,
we might want to not call this exit(1) here.

Reply:
There's a CA ticket to describe the case we met. https://issues.citrite.net/browse/CA-284857.